src/Compiler/Checking/import.fs

```diff
@@ -43,48 +45,101 @@ type AssemblyLoader =
     /// Record a root for a [<Generate>] type to help guide static linking & type relocation
     abstract RecordGeneratedTypeRoot : ProviderGeneratedType -> unit
 #endif
-        
+
+[<Struct; NoComparison>]
+type CanCoerce =
+    | CanCoerce
+    | NoCoerce
+
+type [<Struct; NoComparison; CustomEquality>] TTypeCacheKey =
+
+    val ty1: TType
+    val ty2: TType
+    val canCoerce: CanCoerce
+    val tcGlobals: TcGlobals
+
+    new (ty1, ty2, canCoerce, tcGlobals) =
+        { ty1 = ty1; ty2 = ty2; canCoerce = canCoerce; tcGlobals = tcGlobals }
+
+    interface System.IEquatable<TTypeCacheKey> with
+        member this.Equals other =
+            let ty1 = this.ty1
+            let ty2 = this.ty2
+            let otherTy1 = stripTyEqns this.tcGlobals other.ty1
+            let otherTy2 = stripTyEqns this.tcGlobals other.ty2
+
+            if stampEquals this.tcGlobals ty1 otherTy1 && stampEquals this.tcGlobals ty2 otherTy2 && this.canCoerce = other.canCoerce then
+                true
+            else
+                (
+                    LanguagePrimitives.PhysicalEquality ty1 otherTy1
+                    && LanguagePrimitives.PhysicalEquality ty2 otherTy2
+                    && this.canCoerce = other.canCoerce
+                )
+
+    override this.Equals other =
+        match other with
+        | :? TTypeCacheKey as p -> (this :> System.IEquatable<TTypeCacheKey>).Equals p
+        | _ -> false
+
+    override this.GetHashCode() : int =
+        let g = this.tcGlobals
+
+        let ty1 = stripTyEqns g this.ty1
```

Comment:
Reminder: Deal with repetitive stripping.

Reply:
Refer to previous comment about stripping.